doc/reference/stdio.md

```diff
@@ -0,0 +1,843 @@
+```scheme
+(BufferedReader-read-varint reader (max-bits 64)) -> integer
+```
+Reads a signed integer, with variadic encoding and a maximum integer length of `max-bits`
+Raises an error if the end of input is reached.
+
+#### BufferedReader-read-char
+```scheme
+(BufferedReader-read-char buf) -> char or #!eof
+  buf := BufferedReader
+```
+
+Reads a utf8-encoded character.
+Returns `#!eof` if the endo if input is reached before reading the first byte.
+Raises an error if the end of input is reached for multibyte encoded characters.
+
+#### BufferedReader-read-char!
+```scheme
+(BufferedReader-read-char! buf) -> char
+  buf := BufferedReader
+```
+
+Like `BufferedReader-read-char`, but raises an error if the end of
+input is reached before reading the first byte.
+
+#### BufferedReader-read-string
+```scheme
+(BufferedReader-read-string buf reader str (start 0) (end (string-length str)) (need 0))
+  -> fixnum
+  buf    := BufferedReader
+  str    := string
+  start  := fixnum; 0 <= start < end
+  end    := fixnum; start < end <= (string-length input)
+  need   := fixnum
+```
+
+Reads a string from the buffer; semantics of the `start`, `end` and
+`need` arguments are the same as in `Reader-read`, but applying to
+number of characters.
+Returns the number of characters read.
+If `need > 0` and less than `need` characters are read, an error is raised.
+
+#### BufferedReader-read-line
+```scheme
+(BufferedReader-read-line buf (sep #\newline) (include-sep? #f) (max-chars #f)) -> string
+  buf          := BufferedReader
+  sep          := char or list of chars
+  include-sep? := bool
+  max-chars    := fixnum or #f
+```
+
+Reads a line, separated by `sep` and up to `max-chars` of length.
+The separator is either a single character or a list of characters.
+If `include-sep?` is true, then the separator is include in the stream.
+If the separator is not encountered within `max-chars`, then an error is raised.
+
+#### BufferedReader-skip
+```scheme
+(BufferedReader-skip buf count)
+  buf   := BufferedReader
+  count := fixnum
+```
+
+Skips the next `count` bytes of input.
+If the end of input is encountered before `count` bytes are skipped, an error is raised.
+
+#### BufferedReader-delimit
+```scheme
+(BufferedReader-delimit buf limit) -> BufferedReader
+  buf   := BufferedReader
+  limit := fixnum
+```
+
+returns a new `BufferedReader`, sharing the same buffer, that can read up to `limit` bytes.
+Once the limit is reached, the new buffer signals the end of input.
+
+#### BufferedReader-reset!
+```scheme
+(BufferedReader-reset! buf reader)
+  buf    := BufferedReader
+  reader := Reader
+```
+
+Resets the underlying reader and buffer state, allowing reuse of buffers from a cache.
+
+### BufferedWriter
+The BufferedWriter interface is defined as follows:
+```scheme
+(interface (BufferedWriter Writer)
+  (write-u8 u8)
+  (flush)
+  (reset! output))
+```
+
+The following procedures create and apply to BufferedWriter instances.
```

Comment:
`BufferedWriter`

Reply:
Does the `close` method automatically calls a `flush` on the buffered write, or does the user have explicitly do it first?

Reply:
yes, it flushes.
Will add to the docs.

Reply:
Added documentation about `BufferedWriter-close` flushing the buffer.